ambry-cloud/src/main/java/com/github/ambry/cloud/azure/StorageClient.java

```diff
@@ -72,8 +68,6 @@
  */
 public abstract class StorageClient {
   Logger logger = LoggerFactory.getLogger(StorageClient.class);
-  private AtomicReference<BlobServiceClient> storageClientRef;
```

Comment:
Overall, a better abstraction would be that StorageClient implements methods like `getBlockBlobClient`, `getStorageClient` etc, and hides details about whether these are asyn implementation of sync implementation. It could then define a static method like `buildAsyncStorageClient` or `buildSyncStorageClient` based on a config. All the code everywhere else shouldn't worry about the StorageClient implementation details (whether this is sync implementation or async implemnentation). That way you can swap out any implementation with a new implementation with minimal code changes. It keeps the code extensible.

Reply:
Sure. Thank you for the suggestion. Yeah, like you said, it helps for us to have both sync and async methods. But, since the rest of code like users of StorageClient class is also moving to async, I was just wondering if we would still need sync methods. Also, it seems like the sync azure APIs also use async APIs underneath. Will sync up more with you offline.

Reply:
My main concern here is that a lot of stuff that should be ideally abstracted by StorageClient interface is being leaked out to callers. For example in `AzureBlobDataAccessor::uploadIfNotExists` , the role of `storageClient` should be to upload a blob and maybe return some response. So we should expect a call like `storageClient.uploadWithResponse`. Whether the upload is sync or async, and what goes into making that upload happen shouldn't leak out and be totally handled out by `storageClient`. The contract of `StorageClient` should be -- I can upload a blob to a storage backend (which happens to be azure in our case), I can donwload a blob from a storage backend (which happens to be azure in our case), I can delete a file, I can delete a batch,  I can get some properties for a blob etc. How its does it, is its own business, and callers shouldn't worry about that.

Reply:
Thank you for the explanation. Yeah, it is a good design for abstraction of classes. Like we discussed offline, in this case though we might not need sync implementations of storage client since rest of code is also going to refactored to be asynchronous. For example, we would be changing `AzureBlobDataAccessor::uploadIfNotExists` also to async (i.e. have return type as future which returns immediately and not wait for response) and hence only use storage client async methods.